cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactStore.java

```diff
@@ -282,29 +289,41 @@ private void addAndSortArtifacts(List<ArtifactDetail> artifacts, Row row, int li
       }
 
       ArtifactData data = GSON.fromJson(Bytes.toString(columnEntry.getValue()), ArtifactData.class);
-      sortedArtifacts.put(
-          new ArtifactId(artifactKey.name, new ArtifactVersion(version),
-                         artifactKey.namespace.equals(NamespaceId.SYSTEM) ? ArtifactScope.SYSTEM : ArtifactScope.USER),
-          data);
-    }
-    for (Map.Entry<ArtifactId, ArtifactData> entry : sortedArtifacts.entrySet()) {
-      if (limit != Integer.MAX_VALUE && artifacts.size() == limit) {
-        break;
-      }
-      ArtifactData data = entry.getValue();
-      artifacts.add(new ArtifactDetail(
-        new ArtifactDescriptor(entry.getKey(), Locations.getLocationFromAbsolutePath(locationFactory,
-                                                                                     data.getLocationPath())),
+      ArtifactId artifactId = new ArtifactId(artifactKey.name, new ArtifactVersion(version),
+                                             artifactKey.namespace.equals(NamespaceId.SYSTEM) ?
+                                               ArtifactScope.SYSTEM : ArtifactScope.USER);
+      queue.add(new ArtifactDetail(
+        new ArtifactDescriptor(artifactId, Locations.getLocationFromAbsolutePath(locationFactory,
+                                                                                 data.getLocationPath())),
         data.meta));
+      if (limit != Integer.MAX_VALUE && queue.size() > limit) {
+        queue.poll();
+      }
     }
+    while (!queue.isEmpty()) {
+      artifacts.add(queue.poll());
+    }
+    Collections.reverse(artifacts.subList(0, artifacts.size()));
+  }
+
+  private PriorityQueue<ArtifactDetail> getPriorityQueue(int limit, final ArtifactSortOrder order) {
+    Comparator<ArtifactDetail> comparator = new Comparator<ArtifactDetail>() {
+      @Override
+      public int compare(ArtifactDetail o1, ArtifactDetail o2) {
+        return order == ArtifactSortOrder.ASC ? -o1.getDescriptor().compareTo(o2.getDescriptor()) :
```

Comment:
should use a different Comparator depending on the sort order rather than compare the order every time.